### PR TITLE
Compact HNSW index when ratio of dead bytes / address space is too high

### PR DESCRIPTION
**Pull request by toregge**
relative to used bytes / address space.

This PR is related to issue #18153.

@baldersheim : please review
@geirst : FYI

### CODE REVIEW

**Comment by baldersheim** [comment]
I think it is better if @geirst takes a look at this one.

File: vespalib/src/vespa/vespalib/datastore/array_store.hpp [hunk 1]
```diff
@@ -157,6 +157,20 @@ class CompactionContext : public ICompactionContext {
             }
         }
     }
+    void compact(vespalib::ArrayRef<AtomicEntryRef> refs) override {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/array_store.hpp:160]: Consider if we should have an explicit test for this new function in vespalib/src/tests/datastore/array_store/array_store_test.cpp.

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 2]
```diff
@@ -472,6 +480,89 @@ HnswIndex::trim_hold_lists(generation_t first_used_gen)
...
+    return false;
+}
+
+bool
+HnswIndex::consider_compact_link_arrays(const CompactionStrategy& compaction_strategy)
+{
+    size_t used_bytes = _cached_link_arrays_memory_usage.usedBytes();
+    size_t dead_bytes = _cached_link_arrays_memory_usage.deadBytes();
+    bool compact_memory = compaction_strategy.should_compact_memory(used_bytes, dead_bytes);
+    size_t used_address_space = _cached_level_arrays_address_space_usage.used();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp:528]: Should this be _cached_link_arrays_address_space_usage instead?

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 3]
```diff
@@ -472,6 +480,89 @@ HnswIndex::trim_hold_lists(generation_t first_used_gen)
...
+    size_t dead_address_space = _cached_level_arrays_address_space_usage.dead();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp:529]: Should this be _cached_link_arrays_address_space_usage instead?

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 4]
```diff
@@ -472,6 +480,89 @@ HnswIndex::trim_hold_lists(generation_t first_used_gen)
...
+bool
+HnswIndex::consider_compact_level_arrays(const CompactionStrategy& compaction_strategy)
+{
+    size_t used_bytes = _cached_level_arrays_memory_usage.usedBytes();
+    size_t dead_bytes = _cached_level_arrays_memory_usage.deadBytes();
+    bool compact_memory = compaction_strategy.should_compact_memory(used_bytes, dead_bytes);
+    size_t used_address_space = _cached_level_arrays_address_space_usage.used();
+    size_t dead_address_space = _cached_level_arrays_address_space_usage.dead();
+    bool compact_address_space = compaction_strategy.should_compact_address_space(used_address_space, dead_address_space);
+    if (compact_memory || compact_address_space) {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp:515]: Consider adding a helper function for lines 509-515 to find out if compaction is needed for this aspect. The same helper function can be re-used in consider_compact_link_arrays().

**Review by geirst — APPROVED** [review]
Overall looks good, some comments to consider.

File: vespalib/src/vespa/vespalib/datastore/array_store.hpp (hunk 1 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/datastore/array_store.hpp:160]: Done.

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 2 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Fixed.

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 3 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Fixed

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 4 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp:517]: Done.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**